streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java

```diff
@@ -0,0 +1,293 @@
+        builder
+            .table(INPUT_TOPIC_NAME, Consumed.with(Serdes.Integer(), Serdes.Integer()),
+                Materialized.<Integer, Integer, KeyValueStore<Bytes, byte[]>>as(TABLE_NAME)
+                    .withCachingDisabled())
+            .toStream()
+            .peek((k, v) -> semaphore.release());
+
+        final KafkaStreams client1 = createClient(builder, streamsConfiguration());
+        final KafkaStreams client2 = createClient(builder, streamsConfiguration());
+        final List<KafkaStreams> clients = Arrays.asList(client1, client2);
+
+        final AtomicLong restoreStartOffset = new AtomicLong(-1L);
+        clients.forEach(client -> {
+            client.setGlobalStateRestoreListener(createTrackingRestoreListener(restoreStartOffset));
+            client.start();
+        });
+        waitForClientsToEnterRunningState(clients, 60, TimeUnit.SECONDS);
+
+        produceValueRange(key, 0, batch1NumMessages);
+
+        assertThat("all messages in the first batch were processed in a timely manner",
+            semaphore.tryAcquire(batch1NumMessages, 60, TimeUnit.SECONDS));
+
+        final ReadOnlyKeyValueStore<Integer, Integer> store1 = client1
+            .store(TABLE_NAME, QueryableStoreTypes.keyValueStore());
+
+        final ReadOnlyKeyValueStore<Integer, Integer> store2 = client2
+            .store(TABLE_NAME, QueryableStoreTypes.keyValueStore());
+
+        final boolean client1WasFirstActive;
+        if (store1.get(key) != null) {
+            client1WasFirstActive = true;
+        } else {
+            assertThat("data from the job was sent to the store", store2.get(key) != null);
+            client1WasFirstActive = false;
+        }
+
+        assertThat("no restore has occurred", restoreStartOffset.get(), is(equalTo(-1L)));
+        assertThat("current value in store should reflect all messages being processed",
+            client1WasFirstActive ? store1.get(key) : store2.get(key), is(equalTo(batch1NumMessages - 1)));
+
+        if (client1WasFirstActive) {
+            client1.close();
+        } else {
+            client2.close();
+        }
+
+        final int totalNumMessages = batch1NumMessages + batch2NumMessages;
+
+        produceValueRange(key, batch1NumMessages, totalNumMessages);
+
+        assertThat("all messages in the second batch were processed in a timely manner",
+            semaphore.tryAcquire(batch2NumMessages, 60, TimeUnit.SECONDS));
+
+        assertThat("either restore was unnecessary or we restored from an offset later than 0",
+            restoreStartOffset.get(), is(anyOf(greaterThan(0L), equalTo(-1L))));
+
+        assertThat("current value in store should reflect all messages being processed",
+            client1WasFirstActive ? store2.get(key) : store1.get(key), is(equalTo(totalNumMessages - 1)));
+    }
+
+    private void produceValueRange(final int key, final int start, final int endExclusive) throws ExecutionException, InterruptedException {
```

Comment:
nit: simplify to `throws Exception` -- in test, we don't intent to catch them anyway and just let the test fail